File: Algorithms_C#/95_Unique_Binary_Search_Trees_II.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<TreeNode> GenerateTrees(int n) {
        if (n == 0) {
            return new List<TreeNode>{};
        }
        return generateTrees(1, n);
    }

    public IList<TreeNode> generateTrees(int start, int end){
        if (start > end) {
            return new List<TreeNode>{null};
        }
        IList<TreeNode> allTrees = new List<TreeNode>();
        for (int i = start; i <= end; i++) {
            IList<TreeNode> leftTrees = generateTrees(start, i - 1);
            IList<TreeNode> rightTrees = generateTrees(i + 1, end);

            foreach (TreeNode left in leftTrees) {
                foreach (TreeNode right in rightTrees) {
                    TreeNode currTree = new TreeNode(i);
                    currTree.left = left;
                    currTree.right = right;
                    allTrees.Add(currTree);
                }
            }
        }
        return allTrees;
    }
}

File: Algorithms_C++/95_Unique_Binary_Search_Trees_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> dfs(int start, int end){
        if(start > end){
            return {nullptr};
        }
        vector<TreeNode*> allTrees;

        for(int i = start; i <= end; ++i){
            vector<TreeNode*> leftTrees = dfs(start, i-1);
            vector<TreeNode*> rightTrees = dfs(i+1, end);
            for(auto& left : leftTrees){
                for(auto& right : rightTrees){
                    TreeNode* curTree = new TreeNode(i);
                    curTree->left = left;
                    curTree->right = right;
                    allTrees.push_back(curTree);
                }
            }
        }
        return allTrees;
    }
    vector<TreeNode*> generateTrees(int n) {
        if(!n){
            return {};
        }
        return dfs(1, n);
    }
};